src/invert/laplace/impls/naulin/naulin_laplace.cxx

```diff
@@ -211,46 +242,100 @@ const Field3D LaplaceNaulin::solve(const Field3D &rhs, const Field3D &x0) {
+  int underrelax_count = 0;
+  BoutReal underrelax_factor = initial_underrelax_factor;
+
+  auto calc_b_guess = [&] (const Field3D& x_in) {
+    // Derivatives of x
+    Field3D ddx_x = DDX(x_in, location, DIFF_C2);
+    Field3D ddz_x = DDZ(x_in, location, DIFF_FFT);
+    return rhsOverD - (coords->g11*coef_x_AC*ddx_x + coords->g33*coef_z*ddz_x
+        + coords->g13*(coef_x_AC*ddz_x + coef_z*ddx_x)) - AOverD_AC*x_in;
+  };
+
+  auto calc_b_x_pair = [&, this] (Field3D b, Field3D& x_guess) {
+    // Note take a copy of the 'b' argument, because we want to return a copy of it in the
+    // result
+    // Is it dangerous to take x_guess by reference, if we will call something like
+    // b_x_pair = calc_b_x_pair(b_guess, b_x_pair.second);
+    // ???
+
+    if ( (inner_boundary_flags & INVERT_SET) || (outer_boundary_flags & INVERT_SET) ) {
       // This passes in the boundary conditions from x0's guard cells
-      copy_x_boundaries(x, x0, localmesh);
+      copy_x_boundaries(x_guess, x0, localmesh);
+    }
 
-    // NB need to pass x in case boundary flags require 'x0', even if
-    // delp2solver is not iterative and does not use an initial guess
-    x = delp2solver->solve(b, x);
+    // NB need to pass x_guess in case boundary flags require 'x0', even if delp2solver is
+    // not iterative and does not use an initial guess
+    Field3D x = delp2solver->solve(b, x_guess);
     localmesh->communicate(x);
 
-    // re-calculate the rhs from the new solution
-    // Use here to calculate an error, can also use for the next iteration
-    ddx_x = DDX(x, location, DIFF_C2); // can be used also for the next iteration
-    ddz_x = DDZ(x, location, DIFF_FFT);
-    Field3D bnew = rhsOverD - (coords->g11*coef_x_AC*ddx_x + coords->g33*coef_z*ddz_x + coords->g13*(coef_x_AC*ddz_x + coef_z*ddx_x)) - AOverD_AC*x;
+    return std::pair<Field3D,Field3D>(b, x);
+  };
+
+  Field3D b = calc_b_guess(x0);
+  // Need to make a copy of x0 here to make sure we don't change x0
+  auto b_x_pair = calc_b_x_pair(b, Field3D(x0).allocate());
+  auto b_x_pair_old = b_x_pair;
 
-    Field3D error3D = b - bnew;
+  while (true) {
+    Field3D bnew = calc_b_guess(b_x_pair.second);
+
+    Field3D error3D = b_x_pair.first - bnew;
     error_abs = max(abs(error3D, RGN_NOBNDRY), true, RGN_NOBNDRY);
     error_rel = error_abs / RMS_rhsOverD;
 
-    b = bnew;
+    if (error_rel<rtol or error_abs<atol) break;
 
     count++;
-    if (count>maxits)
-      throw BoutException("LaplaceNaulin error: Took more than maxits=%i iterations to converge.", maxits);
+    if (count>maxits) {
+      throw BoutException("LaplaceNaulin error: Not converged within maxits=%i iterations.", maxits);
+    }
+
+    output<<underrelax_factor<<" "<<count<<" "<<error_abs<<" "<<error_rel<<endl;
+
+    while (error_abs > last_error) {
+      // Iteration seems to be diverging... try underrelaxing and restart
+      underrelax_factor *= .9;
+      underrelax_count++;
```

Comment:
Minor, but `++underrelax_count`

The postfix version increments and returns the old value -- we don't care about the return value, so use prefix instead